velox/type/Filter.cpp

```diff
@@ -170,6 +176,79 @@ bool BigintValuesUsingHashTable::testInt64(int64_t value) const {
   return false;
 }
 
+__m256i BigintValuesUsingHashTable::test4x64(__m256i x) {
+  using V64 = simd::Vectors<int64_t>;
+  auto rangeMask = V64::compareGt(V64::setAll(min_), x) |
+      V64::compareGt(x, V64::setAll(max_));
+  if (V64::compareResult(rangeMask) == V64::kAllTrue) {
+    return V64::setAll(0);
+  }
+  if (containsEmptyMarker_) {
+    return Filter::test4x64(x);
+  }
+  rangeMask ^= -1;
+  auto indices = x * M & sizeMask_;
+  __m256i data = _mm256_mask_i64gather_epi64(
+      V64::setAll(kEmptyMarker),
+      reinterpret_cast<const long long int*>(hashTable_.data()),
+      indices,
+      rangeMask,
+      8);
+  // The lanes with kEmptyMarker missed, the lanes matching x hit and the other
+  // lanes must check next positions.
+
+  auto result = V64::compareEq(x, data);
+  auto missed = V64::compareEq(data, V64::setAll(kEmptyMarker));
+  uint16_t unresolved = V64::compareBitMask(
+      ~V64::compareResult(result) & ~V64::compareResult(missed));
+  if (!unresolved) {
+    return result;
+  }
+  int64_t indicesArray[4];
+  int64_t valuesArray[4];
+  int64_t resultArray[4];
+  *reinterpret_cast<V64::TV*>(indicesArray) = indices + 1;
+  *reinterpret_cast<V64::TV*>(valuesArray) = x;
+  *reinterpret_cast<V64::TV*>(resultArray) = result;
+  auto allEmpty = V64::setAll(kEmptyMarker);
+  while (unresolved) {
+    auto lane = bits::getAndClearLastSetBit(unresolved);
+    // Loop for each unresolved (not hit and
+    // not empty) until finding hit or empty.
+    int64_t index = indicesArray[lane];
+    int64_t value = valuesArray[lane];
+    auto allValue = V64::setAll(value);
+    for (;;) {
+      auto line = V64::load(hashTable_.data() + index);
+
+      if (V64::compareResult(V64::compareEq(line, allValue))) {
+        resultArray[lane] = -1;
+        break;
+      }
+      if (V64::compareResult(V64::compareEq(line, allEmpty))) {
+        resultArray[lane] = 0;
+        break;
+      }
+      index += 4;
+      if (index > sizeMask_) {
+        index = 0;
+      }
+    }
+  }
+  return V64::load(&resultArray);
+}
+
+__m256si BigintValuesUsingHashTable::test8x32(__m256i x) {
```

Comment:
it's not clear to me why the test4x64 code is so much more complicated than the test8x32. Can you add a top level comment to clarify?